usr/src/uts/common/io/drm/drm_sunmod.c

```diff
@@ -263,46 +302,117 @@ drm_gem_map_access(devmap_cookie_t dhp, void *pvt, offset_t offset, size_t len,
+
+	struct drm_gem_object *obj = pvt;
+	struct drm_device *dev = obj->dev;
+	devmap_handle_t *dhp;
+	struct ddi_umem_cookie *cp;
+
+	mutex_enter(&dev->struct_mutex);
+	dhp = (devmap_handle_t *)dhc;
+	cp = (struct ddi_umem_cookie *)dhp->dh_cookie;
+	cp->cook_refcnt++;
+	mutex_exit(&dev->struct_mutex);
+
+	*new_pvtp = obj;
+	return (0);
+}
+
 static void
-drm_gem_unmap(devmap_cookie_t dhp, void *pvtp, offset_t off, size_t len,
-		devmap_cookie_t new_dhp1, void **newpvtp1,
-		devmap_cookie_t new_dhp2, void **newpvtp2)
+drm_gem_unmap(devmap_cookie_t dhc, void *pvt, offset_t off, size_t len,
+		devmap_cookie_t new_dhp1, void **new_pvtp1,
+		devmap_cookie_t new_dhp2, void **new_pvtp2)
 {
+	devmap_handle_t *dhp = (devmap_handle_t *)dhc;
+	devmap_handle_t	*ndhp;
+	struct ddi_umem_cookie *cp;
+	struct ddi_umem_cookie *ncp;
 	struct drm_device *dev;
 	struct drm_gem_object *obj;
 	struct gem_map_list *entry, *temp;
+	boolean_t last_ref = B_FALSE;
 
-	_NOTE(ARGUNUSED(dhp, pvtp, off, len, new_dhp1, newpvtp1))
-	_NOTE(ARGUNUSED(new_dhp2, newpvtp2))
+	_NOTE(ARGUNUSED(off, len))
 
-	obj = (struct drm_gem_object *)pvtp;
+	obj = (struct drm_gem_object *)pvt;
 	if (obj == NULL)
 		return;
 
 	dev = obj->dev;
 
+	/*
+	 * Unload what drm_gem_map_access loaded.
+	 *
+	 * XXX: The devmap_unload() here is probably unnecessary,
+	 * as segdev_unmap() has done a hat_unload() for the
+	 * entire segment by the time we get here.
+	 */
 	mutex_lock(&dev->page_fault_lock);
-	if (list_empty(&obj->seg_list)) {
-		mutex_unlock(&dev->page_fault_lock);
-		return;
+	if (!list_empty(&obj->seg_list)) {
+		list_for_each_entry_safe(entry, temp, struct gem_map_list,
+		    &obj->seg_list, head) {
+			(void) devmap_unload(entry->dhp, entry->mapoffset,
+			    entry->maplen);
+			list_del(&entry->head);
+			drm_free(entry, sizeof (struct gem_map_list), DRM_MEM_MAPS);
+		}
+	}
+	mutex_unlock(&dev->page_fault_lock);
+
+	/*
+	 * Manage ddi_umem_cookie ref counts
+	 */
+
+	mutex_enter(&dev->struct_mutex);
+
+	cp = (struct ddi_umem_cookie *)dhp->dh_cookie;
+	if (new_dhp1 != NULL) {
+		ndhp = (devmap_handle_t *)new_dhp1;
+		ncp = (struct ddi_umem_cookie *)ndhp->dh_cookie;
+		ncp->cook_refcnt++;
+		*new_pvtp1 = obj;
+		ASSERT(ncp == cp);
 	}
 
-	list_for_each_entry_safe(entry, temp, struct gem_map_list,
-	    &obj->seg_list, head) {
-		(void) devmap_unload(entry->dhp, entry->mapoffset,
-		    entry->maplen);
-		list_del(&entry->head);
-		drm_free(entry, sizeof (struct gem_map_list), DRM_MEM_MAPS);
+	if (new_dhp2 != NULL) {
+		ndhp = (devmap_handle_t *)new_dhp2;
+		ncp = (struct ddi_umem_cookie *)ndhp->dh_cookie;
+		ncp->cook_refcnt++;
+		*new_pvtp2 = obj;
+		ASSERT(ncp == cp);
 	}
 
-	mutex_unlock(&dev->page_fault_lock);
+	ASSERT(cp->cook_refcnt > 0);
+	cp->cook_refcnt--;
+	if (cp->cook_refcnt == 0) {
+		last_ref = B_TRUE;
+	}
+
+	mutex_exit(&dev->struct_mutex);
+
+	if (last_ref) {
+		/*
+		 * This is the last mmap reference going away.
+		 * There normally are other references via the
```

Comment:
Confused by the comment, and the `_unreference()` call.  Is there another way to word this?